packages/wallet/ui/src/util/keyManagement.js

```diff
@@ -0,0 +1,425 @@
+    aminoType: 'swingset/WalletAction',
+  },
+});
+
+/**
+ * @typedef {{
+ *   owner: string, // base64 of raw bech32 data
+ *   action: string,
+ * }} WalletAction
+ * @typedef {{
+ *   owner: string, // base64 of raw bech32 data
+ *   spendAction: string,
+ * }} WalletSpendAction
+ */
+
+export const SwingsetRegistry = new Registry([
+  ...defaultRegistryTypes,
+  [SwingsetMsgs.MsgWalletAction.typeUrl, MsgWalletAction],
+  [SwingsetMsgs.MsgWalletSpendAction.typeUrl, MsgWalletSpendAction],
+]);
+
+/**
+ * TODO: estimate fee? use 'auto' fee?
+ * https://github.com/Agoric/agoric-sdk/issues/5888
+ *
+ * @returns {import('@cosmjs/stargate').StdFee}
+ */
+export const zeroFee = () => {
+  const { coinMinimalDenom: denom } = stableCurrency;
+  const fee = {
+    amount: [{ amount: '0', denom }],
+    gas: '100000', // TODO: estimate gas?
+  };
+  return fee;
+};
+
+/** @type {import('@cosmjs/stargate').AminoConverters} */
+const SwingsetConverters = {
+  // TODO: #3628, #4654
+  // '/agoric.swingset.MsgProvision': {
+  //   /* ... */
+  // },
+  [SwingsetMsgs.MsgWalletAction.typeUrl]: {
+    aminoType: SwingsetMsgs.MsgWalletAction.aminoType,
+    toAmino: ({ action, owner }) => ({
+      action,
+      owner: toBech32(bech32Config.bech32PrefixAccAddr, fromBase64(owner)),
+    }),
+    fromAmino: ({ action, owner }) => ({
+      action,
+      owner: toBase64(toAccAddress(owner)),
+    }),
+  },
+  [SwingsetMsgs.MsgWalletSpendAction.typeUrl]: {
+    aminoType: SwingsetMsgs.MsgWalletAction.aminoType,
+    toAmino: ({ spendAction, owner }) => ({
+      spendAction,
+      owner: toBech32(bech32Config.bech32PrefixAccAddr, fromBase64(owner)),
+    }),
+    fromAmino: ({ spendAction, owner }) => ({
+      spendAction,
+      owner: toBase64(toAccAddress(owner)),
+    }),
+  },
+};
+
+/**
+ * key for use in localStorage
+ *
+ * arbitrary; suffix chosen randomly to avoid collisions
+ */
+export const STORAGE_KEY = 'agoric.eis0Aigi';
+
+/**
+ * Maintain a key for signing non-spending messages in localStorage.
+ *
+ * See also `delegateWalletAction()` below.
+ *
+ * @param {object} io
+ * @param {typeof window.localStorage} io.localStorage
+ * @param {typeof import('@cosmjs/crypto').Random.getBytes} io.getBytes for key generation
+ */
+export const makeBackgroundSigner = async ({ localStorage, getBytes }) => {
+  const provideLocalKey = () => {
+    const stored = localStorage.getItem(STORAGE_KEY);
+    if (stored) {
+      return fromBase64(stored);
+    }
+    console.log(
+      `localStorage.setItem(${STORAGE_KEY}, Random.getBytes(${KEY_SIZE}))`,
+    );
+    const seed = getBytes(KEY_SIZE);
+    localStorage.setItem(STORAGE_KEY, toBase64(seed));
+    return seed;
+  };
+  const seed = provideLocalKey();
+  const wallet = await DirectSecp256k1Wallet.fromKey(
+    seed,
+    bech32Config.bech32PrefixAccAddr,
+  );
+
+  const accounts = await wallet.getAccounts();
+  console.debug('device account(s):', accounts);
+
+  const [{ address }] = accounts;
```

Comment:
Use the same `?.` expression here as elsewhere and throw a nicer error?

Reply:
Actually, elsewhere we were dealing with the keplr API, which is doing IO to come up with an address and hence may fail. In this case, we're getting the address from `DirectSecp256k1Wallet`, which is doing only integer calculations that don't have that sort of failure mode. Any failures would show up at the `await wallet.getAccounts()` above.